cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/TimePartitionedFileSetDataset.java

```diff
@@ -539,4 +581,36 @@ public long upgradeLegacyEntries(long startTime, long timeLimitInSeconds) {
     }
   }
 
+  private class BasicTimePartition extends BasicPartition implements TimePartition {
+
+    private final Long time;
+
+    private BasicTimePartition(String relativePath, PartitionKey key) {
+      super(relativePath, key);
+      time = null;
+    }
+
+    private BasicTimePartition(String relativePath, long time) {
+      super(relativePath, partitionKeyForTime(time));
+      this.time = time;
+    }
+
+    @Override
+    public long getTime() {
+      return time != null ? time : timeForPartitionKey(getPartitionKey());
```

Comment:
I am trying to do this in a lazy fashion. In case the caller never calls getTime(), then I don't want to compute it. I could actually set it here if it is null.